ambry-router/src/main/java/com.github.ambry.router/GetBlobInfoOperation.java

```diff
@@ -335,27 +337,32 @@ private void handleBody(InputStream payload, MessageMetadata messageMetadata)
       // submit decrypt job
       progressTracker.initializeDecryptionTracker();
       logger.trace("Submitting decrypt job for {}", blobId);
+      decryptJobMetricsTracker = new CryptoJobMetricsTracker(routerMetrics.decryptJobMetrics);
+      decryptJobMetricsTracker.startTracker();
       long startTimeMs = System.currentTimeMillis();
       cryptoJobHandler.submitJob(
           new DecryptJob(blobId, encryptionKey.duplicate(), null, userMetadata, cryptoService, kms,
-              (DecryptJob.DecryptJobResult result, Exception exception) -> {
-                logger.trace("Handling decrypt job callback results for {}", blobId);
-                routerMetrics.decryptTimeMs.update(System.currentTimeMillis() - startTimeMs);
-                if (exception == null) {
-                  logger.trace("Successfully updating decrypt job callback results for {}", blobId);
-                  operationResult = new GetBlobResultInternal(
-                      new GetBlobResult(new BlobInfo(serverBlobProperties, result.getDecryptedUserMetadata().array()),
-                          null), null);
-                  progressTracker.setDecryptionSuccess();
-                } else {
-                  logger.trace("Exception {} thrown on decryption for {}", exception, blobId);
-                  setOperationException(
-                      new RouterException("Exception thrown on decrypting the content for " + blobId, exception,
-                          RouterErrorCode.UnexpectedInternalError));
-                  progressTracker.setDecryptionFailed();
-                }
-                routerCallback.onPollReady();
-              }));
+              decryptJobMetricsTracker, (DecryptJob.DecryptJobResult result, Exception exception) -> {
+            decryptJobMetricsTracker.startResponseProcessing();
+            logger.trace("Handling decrypt job callback results for {}", blobId);
+            routerMetrics.decryptTimeMs.update(System.currentTimeMillis() - startTimeMs);
+            if (exception == null) {
+              logger.trace("Successfully updating decrypt job callback results for {}", blobId);
+              operationResult = new GetBlobResultInternal(
+                  new GetBlobResult(new BlobInfo(serverBlobProperties, result.getDecryptedUserMetadata().array()),
+                      null), null);
+              progressTracker.setDecryptionSuccess();
+            } else {
+              decryptJobMetricsTracker.incrementOperationError();
+              logger.trace("Exception {} thrown on decryption for {}", exception, blobId);
+              setOperationException(
+                  new RouterException("Exception thrown on decrypting the content for " + blobId, exception,
+                      RouterErrorCode.UnexpectedInternalError));
+              progressTracker.setDecryptionFailed();
+            }
+            decryptJobMetricsTracker.completeResponseProcessing();
```

Comment:
shouldn't this be after the `onPollReady()`? That's when the processing is truly complete

Reply:
I felt it doesn't make much of a diff. Will change it.